catalyst/contrib/data/ml/query_gallery_dataset.py

```diff
@@ -0,0 +1,148 @@
+from typing import Dict
+from abc import ABC, abstractmethod
+import os
+from pathlib import Path
+
+from PIL import Image
+
+import torch
+from torch.utils.data import Dataset
+from torchvision import transforms
+
+
+class QueryGalleryDataset(Dataset, ABC):
+    """
+    QueryGallleryDataset for CMCScoreCallback
+    """
+
+    @abstractmethod
+    def __getitem__(self, item) -> Dict[str, torch.Tensor]:
+        """
+        Dataset for query/gallery split should
+        return dict with `embeddings`, `labels` and
+        `is_query` key. Value by key `is_query` should
+        be boolean and indicate whether current object
+        is in query or in gallery.
+        Raises:
+            NotImplementedError:
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def query_size(self) -> int:
+        """
+        Query/Gallery dataset should have property
+        query size
+        Returns:
+            query size
+        Raises:
+            NotImplementedError:
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def gallery_size(self) -> int:
+        """
+        Query/Gallery dataset should have property
+        gallery size
+        Returns:
+            gallery size
+        Raises:
+            NotImplementedError:
+        """
+        raise NotImplementedError
+
+
+class QueryGalleryFolderDataset(QueryGalleryDataset):
+    """
+    QueryGalleryFolderDataset
+
+    Torchvision style dataset for query gallery data.
+    Expecting input: ::
+
+        path/query/class_x/xxx.ext
+        path/query/class_y/xxy.ext
+        path/query/class_z/xxz.ext
+
+        path/gallery/class_x/123.ext
+        path/gallery/class_y/nsdf3.ext
+        path/gallery/class_z/asd932_.ext
+    """
+
+    def __init__(self, path: str, transform=None):
+        """
+
+        Args:
+            path: path to dataset
+            transform: transforms for images
+        """
+        path = Path(path)
+        query_path = path / "query"
+        gallery_path = path / "gallery"
+        self.query_labels = []
+        self.query_imgs = []
+        for label in os.listdir(query_path):
+            img_path = query_path / str(label)
+            for img in os.listdir(img_path):
+                pil_image = Image.open(img_path / img)
+                self.query_imgs.append(pil_image)
+                self.query_labels.append(int(label))
+        self.gallery_labels = []
+        self.gallery_imgs = []
+        for label in os.listdir(gallery_path):
+            img_path = query_path / str(label)
+            for img in os.listdir(img_path):
+                pil_image = Image.open(img_path / img)
+                self.gallery_imgs.append(pil_image)
+                self.gallery_labels.append(int(label))
```

Comment:
you dont need to keep them in memory, read them at `get`

Reply:
btw, you can use https://github.com/catalyst-team/catalyst/blob/master/catalyst/contrib/utils/dataset.py#L13